### PR TITLE
Initial draft to check CA subject DN in X.509 client authenticator

### LINKED ISSUE #48412: Add option to specify the CA Subject DN in the Client X.509 Authenticator
### Description

Right now the X.509 client authenticator only allows to specify the Subject DN for the client certificate. As the trust-store (or mTLS store) is common for all the keycloak setup (same for all the realms), we can enhance this to also add the Root CA Subject DN. This way keycloak can check if the client certificate is issued by the expected CA. A very raw idea of the implementation:

* We need to re-verify the certificate in the app side and obtain the root CA (using the keycloak trust-store).
* If trusted (it should be as it was previously verified by the mTLS, but it can be more complicated when using proxies), check the Root CA has the correct and configured Subject DN.
* Check how the certificate chain is obtained in proxies. We need to obtain the full chain passed by the client.
* Document this and add a note for adding intermediate CAs if the chain is not passed by proxies.
* Add an option to disable the new working or just maintain the same old behavior if the CA DN is not specified but make it compulsory. Secure by default but old configurations maintain the old behavior. We can think about this when doing the implementation.

For the moment this issue is just for the client X.509 authentication. We will need a follow up for the user authenticator (but in that case we will need to setup a list of valid CA DNs, as maybe there are several valid CAs).

### PR DESCRIPTION
**Pull request by rmartinc**
Closes #48412

Initial draft to also check the CA Subject DN in the client X.509 authenticator. The main ideas are the following:

* There is a new multi attribute to specify the Subject DN for the CA in the client authenticator. It is multi because when the chain is not complete you can configure the intermediate CA too (the chain is trusted until the intermediate CA that should be included in the truststore).
* The truststore is read in keycloak using the `TruststoreBuilder`. This part is nasty and I don't like it (I uses `System.setProperty` following what it's done today for the common TLS properties). We can improve it, I accept ideas (maybe creating a truststore directly in the builder? But for the moment I want to receive some feedback for the whole PR).
* The new attribute is required for the X.509 authenticator. I modified the JS to force the Subject DN and the CA Subject DN. It is also validated in the client validator at creation/update time in the server. So if you modify the client you need to configure both attributes. Old clients work as before. This is the way I'm trying to manage backwards compatibility (we cannot update anything here, as the CA subject DN is something we cannot know in the upgrade). We can be less restrictive and just do it in the JS part (not in the validator), if you think that it is enough.
* There is another issue here. The TODO in the `X509ClientAuthenticator` class. Currently the class is using `certificate.getSubjectDN().getName()` to match the regex (when using regex), which is deprecated and in the javadoc mentions that is internal implementation that should not be used (I have even checked that the result is different for Sun or BCFIPS implementation for example). I don't know what to do here, it only affects when the regular expression is used. I would say that we need to use RFC2253 (`getX500Principal().getName(X500Principal.RFC2253, CUSTOM_OIDS)`). But this is totally a breaking change. And we cannot adapt the regexp at upgrade or similar. For the moment I'm maintaining the same `getSubjectDN`, but it's incorrect for sure. Maybe we can just change and add a switch in the factory to revert to previous (incorrect) behavior.
* Test updated to new TS, and the `ManagedCertificates` helper now allows you to configure stores directly (instead of creating them). This way we can add complex configurations with CA, intermediate,... I'm re-using the certificates from the old TS. I have not removed the old test in the draft.

@stianst @mposolda @ahus1 @vmuzikar @shawkins WDYT?

### CODE REVIEW

File: services/src/main/java/org/keycloak/truststore/FileTruststoreProviderFactory.java [hunk 1]
```diff
@@ -129,7 +129,7 @@ public void init(Config.Scope config) {
 
         TruststoreCertificatesLoader certsLoader = new TruststoreCertificatesLoader(truststore);
         provider = new FileTruststoreProvider(truststore, verificationPolicy, Collections.unmodifiableMap(certsLoader.trustedRootCerts)
-                , Collections.unmodifiableMap(certsLoader.intermediateCerts));
+                , Collections.unmodifiableMap(certsLoader.intermediateCerts), TruststoreBuilder.getHttpsTruststore());
```
Comment (shawkins) [inline services/src/main/java/org/keycloak/truststore/FileTruststoreProviderFactory.java]: If there's really no build time work that needs captured, instead of the static / system properties on the TruststoreBuilder, you can directly use calls like `config.root().get(HttpOptions.HTTPS_TRUST_STORE_FILE.getKey())` here.

**Comment by ahus1** [comment]
> when the chain is not complete you can configure the intermediate CA too (the chain is trusted until the intermediate CA that should be included in the truststore)

IMHO this is not how it should work to validate certificates. 

The standard way is: 

* The client provides its client certificate and possible intermediate certificates that lead up to the root certificate. 
* The server builds the full trust chain until it reaches the trusted root certificate. 
* If the server can't build a full chain of trust, the validation basically failed. 

Yes, the server could "help" by providing those intermediate certificates, but that's not their job, but maybe that's ok. 

If you would allow an intermediate by just looking at its common name, then any other CA could present such an intermediate CA common name, and then you would be IMHO vulnerable. So I'd vote for not configuring any intermediate CAs.

**Comment by rmartinc** [comment]
@ahus1 Argh! Yes, I defaulted again to normal certificate thinking. And yes, that is possible, using a CertPathBuilderResult. I just need to separate trust and intermediate certs. Thanks for spotting this! OK, I'm going to change that and set just one CA Subject DN for the authenticator. I'm going to try what @shawkins comments.

I also decided what to do with the subject DN check. I think that I'm going to deprecate the regex switch option and the CA Subject DN is only checked with full DN. The regex makes no sense.

**Comment by ahus1** [comment]
> Thanks for spotting this!

Please stop me if you think I'm too strict. 

+1 for not having a regex for the CA Subject DN. 

About multiple CAs: Thinking about this more, a setup might have different CAs that it trusts, with a different set of user CNs/rules what to accept. For such a setup, would I then have multiple authenticators that are available as alternatives? When for such a setup multiple authenticators are used, only a single DN for the CA would be necessary.

**Comment by rmartinc** [comment]
For the moment this is just for the X.509 client authenticator. It's big enough to not add more things. For the client I think that we don't need multiple CAs, just one cert for the client, so one CA. For the users I would say that maybe we need more than one, maybe they have different CAs valid for users, so users can present different valid certificates but the data to match (for example email) is always in the certs. But this is more for the next PR.

**Comment by ahus1** [comment]
@rmartinc - Hm, might still be the same for the client CAs. Can I combine multiple client authenticators in that flow?

**Comment by ahus1** [comment]
@tsaarni - I got a question for you as you implemented "Authorizing clients that are allowed to send XFCC headers" for the envoy proxy setup: https://github.com/Nordix/keycloak-client-cert-lookup-for-envoy

I see that you added an option `spi-x509cert-lookup-envoy-cert-path-verify` that is a path, so it includes the leaf subject CN, intermediate subjects, and eventually the root CA. 

With this issue at hand, I wonder if you would consider specifying the lead subject CN plus the root CA CN of similar ease of use and security. 

Or in other words, when authenticating clients, is Subject CN + Root CN the same as Subject CN + Intermediate CN + Root CN? Would you be missing a feature if it would not allow Subject CN + Intermediate CN?

Thanks!

**Comment by rmartinc** [comment]
@ahus1 In theory the certificate chain sent to the server is all the cert chain minus one (the CA is not present). But sometimes only the initial cert is sent by the client, I have always understood that this was a mistake, but it happens. In order to be able to verify the cert in this situation (missing intermediate), keycloak should also have the intermediate in the truststore to complete the validation (it's more or less the same for the TLS part). There is a test that is doing exactly this, so we are already testing that a single cert (missing intermediate) is working. I'm finishing the change to have a second round of comments.

**Comment by tsaarni** [comment]
While root CAs are more commonly used as trust anchors, intermediate CAs can also be configured for partial chain validation. I think this is generally a convenience feature rather than a necessity.

In my client certificate lookup provider project, I allowed the combinations (leaf alone, leaf plus intermediates, or the full chain) just for flexibility. I guess one could have used partial chain validation if wanted (we don't).

If the check supported only the leaf and root, then edge cases like below could not be expressed unambiguously.

image

**Comment by rmartinc** [comment]
Okis, tests passed OK. Now the changes are:

* The ca subject DN always uses a exact DN. It's a single attribute in the configuration for the client authenticator.
* The certificate is reconstructed with a CertPathBuilder like it was done for users previously. So we always reach the trust anchor (if there are missing certs, the keycloak admin should add the needed intermediate certs in the trustore).
* I have declared the switch attribute for regex attribute deprecated because it's problematic. It's only used as before for the subejct DN of the certificate, not for the new ca attribute.

There are missing things like upgrading notes, better labels and things like that. But I think the functionality is complete. Probably I will add another change to not use the https truststore if the `proxy-headers` option is passed (I think that in that case using normal trust-store is better than the specific one). Please take a look to see if I'm missing something. Adding the new tests for FIPS can be also a long step. So if you see this OK, I will try to polish this to have everything in place (notes, labels, FIPS testsuite,...).

**Comment by rmartinc** [comment]
I think that now it's complete. I have added the notes for 26.7. Don't know if we want this to be backported but probably we better maintain it for 26.7. Please take a look when you have time.

**Comment by ahus1** [comment]
@rmartinc - this looks good to me, thank you for the change. I reviewed the docs, and it makes sense to include this only in 26.7 as it is a breaking change (though for good reasons). 

I also see that it also logs warnings. 

The only real breaking change is IMHO that the import would fail if the information is missing. Could we be more gratuitous here? Maybe make this field only mandatory in the UI, and add the breaking changes only in 27?
I know that some people use the import to setup their staging environments, and they might be a bit annoyed with a breaking change. We're doing for the right reasons, but maybe it is ok to move the breaking part to 27.

Some nitpicks: 

* In the migration guide, there is a separate section for deprecations, it might be worth repeating the deprecated part there. 
* There is a deprecation warning in the log when it is used. It would be good to log the realm ID and the clientId so people would know where to fix this. 
* When adding deprecations, add a in the message or at least in the code a comment when it was deprecated. This helps later to eventually remove the code.   
* The warning when there is no CA specified should also state that this is deprecated, not only that it enhances security.

File: test-framework/core/src/main/java/org/keycloak/testframework/https/ManagedCertificates.java [hunk 2]
```diff
@@ -119,6 +131,27 @@ public SSLContext getClientSSLContext() {
...
+        KeyStore truststoreStore = load(truststorePath);
+        if (mTlsEnabled) {
+            Path keystorePath = mTlsEnabled ? checkPath(keystore) : null;
+            KeyStore keystoreStore = load(keystorePath);
+            return createClientSSLContext(keystoreStore, truststoreStore, mTlsEnabled);
+        } else {
+            return createClientSSLContext(null, truststoreStore, false);
+        }
+    }
+
```
Comment (vaceksimon) [inline test-framework/core/src/main/java/org/keycloak/testframework/https/ManagedCertificates.java:154]: This class should only return an SSLContext created from the KeyStore and TrustStore loaded at server startup. If a test needs a different SSL context for testing purposes, I think the responsibility should be up to them to create it.

File: test-framework/core/src/main/java/org/keycloak/testframework/https/CertificatesConfigBuilder.java [hunk 3]
```diff
@@ -55,4 +61,37 @@ public CertificatesConfigBuilder mTlsEnabled(boolean mTlsEnabled) {
...
+     * @param serverTruststore
+     * @param clientKeystore
+     * @param clientTruststore
+     * @return this
+     */
+    public CertificatesConfigBuilder stores(String serverKeystore, String serverTruststore, String clientKeystore, String clientTruststore) {
+        this.serverKeystore = ManagedCertificates.checkPath(serverKeystore);
+        this.serverTruststore = ManagedCertificates.checkPath(serverTruststore);
+        this.clientKeystore = ManagedCertificates.checkPath(clientKeystore);
+        this.clientTruststore = ManagedCertificates.checkPath(clientTruststore);
```
Comment (vaceksimon) [inline test-framework/core/src/main/java/org/keycloak/testframework/https/CertificatesConfigBuilder.java]: The conversion from `String` to `Path` and any checks shouldn't be in a configuration class. Putting these four lines in the `ManagedCertificates` class is a better option here.

File: test-framework/core/src/main/java/org/keycloak/testframework/https/CertificatesConfigBuilder.java [hunk 4]
```diff
@@ -55,4 +61,37 @@ public CertificatesConfigBuilder mTlsEnabled(boolean mTlsEnabled) {
...
+    /**
+     * Configure manually the stores using files in the classpath.
+     *
+     * @param serverKeystore
+     * @param serverTruststore
+     * @param clientKeystore
+     * @param clientTruststore
+     * @return this
+     */
+    public CertificatesConfigBuilder stores(String serverKeystore, String serverTruststore, String clientKeystore, String clientTruststore) {
```
Comment (vaceksimon) [inline test-framework/core/src/main/java/org/keycloak/testframework/https/CertificatesConfigBuilder.java:89]: Since this is a builder, it begs the question of whether this method should be split into 4 methods for each store. But all four arguments need to be set anyway, so it wouldn't change much.

**Review by vaceksimon — CHANGES_REQUESTED** [review]

File: test-framework/core/src/main/java/org/keycloak/testframework/https/CertificatesConfigBuilder.java (hunk 3 above)
Comment (rmartinc) [reply test-framework/core/src/main/java/org/keycloak/testframework/https/CertificatesConfigBuilder.java]: Done!

File: test-framework/core/src/main/java/org/keycloak/testframework/https/CertificatesConfigBuilder.java (hunk 4 above)
Comment (rmartinc) [reply test-framework/core/src/main/java/org/keycloak/testframework/https/CertificatesConfigBuilder.java:89]: Yep, my idea was showing all the `stores` are needed. If you want to separate them in different methods is OK to me, but I preferred this way to show all of them are necessary.

File: test-framework/core/src/main/java/org/keycloak/testframework/https/ManagedCertificates.java (hunk 2 above)
Comment (rmartinc) [reply test-framework/core/src/main/java/org/keycloak/testframework/https/ManagedCertificates.java:154]: I have not done this point for the moment. Maybe we can create a KeystoreUtils or similar and this class and the test class use the utility methods. But IMO the test that needs to use different client certificates is going to always inject the certificates, so here is enough. If you really prefer to use a utility class just let me know.

**Comment by rmartinc** [comment]
> The only real breaking change is IMHO that the import would fail if the information is missing. Could we be more gratuitous here? Maybe make this field only mandatory in the UI, and add the breaking changes only in 27? I know that some people use the import to setup their staging environments, and they might be a bit annoyed with a breaking change. We're doing for the right reasons, but maybe it is ok to move the breaking part to 27.

It makes sense, but then the kc.sh or direct endpoints can be used to set a configuration without the CA subject DN. For me it's OK and just doing the check in the console is enough (but I wanted to be extra secure). I have added the change. Thinking now maybe we can also add the warning in the validation but let it goes through. For the moment I have just commented the previous validation. I wanted to see the tests. I tested locally and it seems to work OK but I want to be sure. I'll add the warning tomorrow if I see the CI passes OK.

> Some nitpicks:
> 
>     * In the migration guide, there is a separate section for deprecations, it might be worth repeating the deprecated part there.

Done.

>     * There is a deprecation warning in the log when it is used. It would be good to log the realm ID and the clientId so people would know where to fix this.
>     * When adding deprecations, add a in the message or at least in the code a comment when it was deprecated. This helps later to eventually remove the code.

Done in both, clientId and realm name.

>     * The warning when there is no CA specified should also state that this is deprecated, not only that it enhances security.

Done.

**Comment by rmartinc** [comment]
rebasing

**Review by vaceksimon — APPROVED** [review]
Approving the test-related changes

**Review by lhanusov — APPROVED** [review]

**Comment by rmartinc** [comment]
@ahus1 Please check if the changes address your suggestions...

**Review by ahus1 — DISMISSED** [review]
Thank you, my comments have been addressed. Approving but not merging yet, as I assume someone from your team wants  to review it as well.

**Comment by rmartinc** [comment]
@ahus1 I think that @vaceksimon has already approved it from core-authn team, so you can go with this one.

**Review by edewit — DISMISSED** [review]

**Comment by rmartinc** [comment]
Rebased because of conflicts.

**Review by mposolda — COMMENTED** [review]
@rmartinc The changes makes sense to me, Thanks for the PR!

Approving, but did not review deeply whole implementation as I rely on other reviews of this PR.

Just some points:

1) This PR will possibly cause some issues with dynamic client registration as the metadata mentioned in the [OAuth 2 MTLS client authentication RFC](https://datatracker.ietf.org/doc/html/rfc8705#section-2.1.2) does not have anything like `Certificate authority subject DN` or similar. So the dynamically registered clients, which rely on the MTLS likely will not work OOTB. Maybe next step is to make client-policy executor, which will configure the `Certificate authority subject DN` for the dynamically registered clients. But that is for a follow-up though?

2) Another point (slightly related to the above): When the option `Certificate authority subject DN`  is configured at the client level, it means that client administrator has permission to configure it. So the client administrator would have permission to configure it to the possibly untrusted CA. I can think of some example scenario like:
  - Server has 2 certificate authorities in truststore `ca=myroot1` and `ca=myroot2`. But realm admin of realm `foo` wants to trust only certificates issued by `ca=myroot1`
  - Client administrator of client `myclient` in realm `foo` has certificate with subject `ca=mysubject` signed by `ca=myroot2` (which is unexpected CA as realm admin wants just `ca=myroot1`) . But client administrator will be able to configure `Certificate authority subject DN` to `ca=myroot2` and hence bypass the requirements by realm admin.

Maybe this is a corner-case though? At the same time, the assumption is that client administrators are not so trusted guys and their permission should be limited... However maybe the client policies are the solution to this? The client-policy executor I've mentioned in point 1 can be possibly used to also enforce that `Certificate authority subject DN` is as expected.

I am approving as these are likely corner-cases and this PR improves the situation. But just something to consider :-)

**Review by mposolda — DISMISSED** [review]

**Comment by rmartinc** [comment]
Thanks @mposolda!

I think that we need to add something for point 1, at least the new attribute should be added to the representation and be passed to the endpoint. I'll update the upgrading notes with this point saying that now two attributes should be passed to configure X509 authentication vie registration endpoint.

For point 2 I think that we are OK. If the client admin change the jwks URL is exactly the same for JWT authetnication, they can obviously change how the client authenticates. I see the X509 CA the same. If the client admin wants to do bad things to authenticate client, he can do it (note he can even change the client to use a secret and use it directly). The problem here is someone else different to the admin uses another CA for another realm for example.

So wait for the moment, I will add changes for point 1.

**Comment by mposolda** [comment]
@rmartinc The problem is that OIDC client representation does not have anything like `Certificate authority subject DN` as a metadata attribute . So I am not sure how to pass it with client registration request in specs compliant way?

If you want to address this, I see the possibility is either have the `Certificate authority subject DN` either at the realm level or some client-policy executor, which will set the default as long as it is executed (that will allow more flexibility with different clients of same realm using different `Certificate authority subject DN`, but client policies are more complicated to setup). But maybe I am missing some other option?

**Comment by rmartinc** [comment]
Yep, that's the problem of doing weird things with certificates. :smile:

For the moment we are not enforcing the attribute, just in the admin console. So it will work. So we have to think about that problem in the future (if we enforce it).

**Comment by rmartinc** [comment]
I think we can add a client registration policy to add a default CA subject DN. I will try that too.

**Comment by mposolda** [comment]
Yes, client registration policy is also an option. I was rather thinking the "client policy" (as client registration policies are maybe going to be deprecated/removed at some point in favour of client policies, but not sure when it happens...). But client registration policy is also OK

**Comment by rmartinc** [comment]
@mposolda I have added a client policy to add a default CA subject DN for the client registration. A custom medatadata attribute is created too. I added a extra boolean property  to reject any other CA (this limits the client admins too). Changes in the last commit. WDYT?

File: core/src/main/java/org/keycloak/representations/oidc/OIDCClientRepresentation.java [hunk 5]
```diff
@@ -103,6 +103,9 @@ public class OIDCClientRepresentation {
 
     private String tls_client_auth_subject_dn;
 
+    // custom attribute to set the CA subject DN in X509 authentication
+    private String tls_client_auth_ca_subject_dn;
```
Comment (mposolda) [inline core/src/main/java/org/keycloak/representations/oidc/OIDCClientRepresentation.java]: My vote is to go simple in this PR and not introduce the custom attribute and just always make executor to assign configured value.

The parties, which use OIDC dynamic client registration usually need to be specs compliant and using custom (KC specific) attributes does not work for those. We do not use the custom attribute for anything else in the `OIDCClientRepresentation`. Moreover, it is not documented (besides `changes-26_7_0.adoc`, but those are just upgrading instruction, but not a proper docs).

**Review by mposolda — COMMENTED** [review]
@rmartinc Added a review of last commit.

File: core/src/main/java/org/keycloak/representations/oidc/OIDCClientRepresentation.java (hunk 5 above)
Comment (rmartinc) [reply core/src/main/java/org/keycloak/representations/oidc/OIDCClientRepresentation.java]: Done!

**Review by mposolda — APPROVED** [review]
@rmartinc Thanks!
